### ISSUE TITLE
Allow shorthand vector format for compatibility with Spark/ES connector

### ISSUE THREAD

**Issue by alexklibisz** [alexklibisz/elastiknn#177]
Have heard from a couple people via email that they'd like to use Elastiknn with the spark elasticsearch connector.
I've personally never used it, but it seems like a reasonable request. I'm guessing the main usecase for this is indexing vectors. 
Ideally, we'd end up with an example in the examples directory that demonstrates how to use that connector.

It would be great if someone can post some example code or a few sentences showing/describing how they'd like to use the connector.

**Comment by divdasani** [comment]
The spark connector does not work with nesting. The connector treats the column name of a dataframe as a key and the row value as the value in the `JSON`. Thus, the spark connector should work seamlessly with ElastiKNN if the `JSON` structure for indexing documents has no nesting. ie modifying the JSON structure of documents with an `elastiknn_dense_float_vector` field from
```
POST /my-index/_doc
{
    "my_vec": {
        "values": [0.1, 0.2, 0.3, ...] 
    }
}
```
to
```
POST /my-index/_doc
{
    "my_vec": [0.1, 0.2, 0.3, ...]
}
```
should resolve the issue.

> But what happens if the vector is nested further in the document, e.g., `user.user_vec.values`?
For this reason, I suspect most people who utilize the spark connector do not use nested fields in their documents.

**Comment by alexklibisz** [comment]
I'm a bit stumped on how to solve this at the moment.
Specifically, I don't see a way to support this sort of shorthand vector syntax for both dense float and sparse bool vectors.
It's fairly simple to implement it so that `"my_vec": [0.1, 0.2, ...]` is parsed as a valid dense float vector.
It's less obvious how to do this for sparse bool vectors, since they really require two fields: the total number of indices, and the list of `true` indices. If you leave off the total number of indices and just take a list of values, then the dense and sparse shorthand will be almost identical, distinguished only in that one of them is an array of floats and the other an array of ints. But it's super easy to mix up floats and ints anyways, especially in a language like python where you're not strictly defining types of list/array elements.

I was reading this issue: https://github.com/elastic/elasticsearch-hadoop/issues/948, and it looks like nested documents _might_ actually be supported? Similarly here: https://stackoverflow.com/a/33549964/1695454

I'll see if I can spin up a local spark environment and try indexing some random vectors.

**Comment by divdasani** [comment]
It looks like the first solution you posted explains how to convert `JSON` data to a dataframe; we have the opposite requirement here. The seconds solution seems more promising, but requires that data in the dataframe is _already_ preformatted as a `JSON`, which will be an inconvenient ask of users.

Given that you only have these two vector data type fields (`elastiknn_sparse_bool_vector` and `elastiknn_dense_float_vector`), one recommendation I suggest if you want to maintain clarity between the two fields is to have `elastiknn_sparse_bool_vector` accept a list encompass the two values you specify above. In other words, modify the `JSON` structure from:
```
POST /my-index/_doc
{
    "my_vec": {
       "true_indices": [1, 3, 5, ...],
       "total_indices": 100,             
    }
}
```
to:
```
POST /my-index/_doc
{
    "my_config": [[1, 3, 5, ...], 100]
}
```
Though the implementation may not be the cleanest, it will be far more convenient to users and is acceptable given there are only two vector data types.

**Comment by alexklibisz** [comment]
I see. Yeah that could work.

**Comment by alexklibisz** [comment]
Started to look into this yesterday (#179 ). ES makes it a bit trickier than expected. It treats the original format as an object which can be easily converted to a java map and then parsed into a vector. It treats the shorthand as a sort-of stream of floats with unknown length. So you can't just say, "try the original format, and if that fails, try the shorthand format". It should be doable, but will have to wait til at least tomorrow.

**Comment by alexklibisz** [comment]
@divdasani When you get a chance, try out this release: https://github.com/alexklibisz/elastiknn/releases/tag/0.1.0-PRE43-PR179-SNAPSHOT

I implemented the shorthand format as you described above. All implemented and described again in this PR: https://github.com/alexklibisz/elastiknn/pull/179

LMK how that works for you. I think I like what we arrived at. It will probably make the python client a bit simpler as well.
